elliptic-curve/src/point/non_identity.rs

```diff
@@ -103,6 +107,79 @@ impl<P> AsRef<P> for NonIdentity<P> {
     }
 }
 
+impl<const N: usize, P> BatchNormalize<[Self; N]> for NonIdentity<P>
+where
+    P: Curve + BatchNormalize<[P; N], Output = [P::AffineRepr; N]>,
+{
+    type Output = [NonIdentity<P::AffineRepr>; N];
+
+    fn batch_normalize(points: &[Self; N]) -> [NonIdentity<P::AffineRepr>; N] {
+        // Ensure casting is safe.
+        // This always succeeds because `NonIdentity` is `repr(transparent)`.
+        debug_assert_eq!(size_of::<P>(), size_of::<NonIdentity<P>>());
+        debug_assert_eq!(align_of::<P>(), align_of::<NonIdentity<P>>());
+
+        #[allow(unsafe_code)]
+        // SAFETY: `NonIdentity` is `repr(transparent)`.
+        let points: &[P; N] = unsafe { &*points.as_ptr().cast() };
+        let affine_points = <P as BatchNormalize<_>>::batch_normalize(points);
+
+        // Ensure `array::map()` can be optimized to a `memcpy`.
+        debug_assert_eq!(
+            size_of::<P::AffineRepr>(),
+            size_of::<NonIdentity<P::AffineRepr>>()
+        );
+        debug_assert_eq!(
+            align_of::<P::AffineRepr>(),
+            align_of::<NonIdentity<P::AffineRepr>>()
+        );
+
+        affine_points.map(|point| NonIdentity { point })
+    }
+}
+
+#[cfg(feature = "alloc")]
+impl<P> BatchNormalize<[Self]> for NonIdentity<P>
+where
+    P: Curve + BatchNormalize<[P], Output = Vec<P::AffineRepr>>,
+{
+    type Output = Vec<NonIdentity<P::AffineRepr>>;
+
+    fn batch_normalize(points: &[Self]) -> Vec<NonIdentity<P::AffineRepr>> {
+        // Ensure casting is safe.
+        // This always succeeds because `NonIdentity` is `repr(transparent)`.
+        debug_assert_eq!(size_of::<P>(), size_of::<NonIdentity<P>>());
+        debug_assert_eq!(align_of::<P>(), align_of::<NonIdentity<P>>());
+
+        #[allow(unsafe_code)]
+        // SAFETY: `NonIdentity` is `repr(transparent)`.
+        let points: &[P] = unsafe { &*(points as *const [NonIdentity<P>] as *const [P]) };
+        let affine_points = <P as BatchNormalize<_>>::batch_normalize(points);
+
+        // Ensure casting is safe.
+        // This always succeeds because `NonIdentity` is `repr(transparent)`.
+        debug_assert_eq!(
+            size_of::<P::AffineRepr>(),
+            size_of::<NonIdentity<P::AffineRepr>>()
+        );
+        debug_assert_eq!(
+            align_of::<P::AffineRepr>(),
+            align_of::<NonIdentity<P::AffineRepr>>()
+        );
+
+        #[allow(unsafe_code)]
+        // SAFETY: `NonIdentity` is `repr(transparent)`.
+        let result: Vec<NonIdentity<P::AffineRepr>> = unsafe {
+            (&affine_points as *const Vec<P::AffineRepr>)
+                .cast::<Vec<NonIdentity<P::AffineRepr>>>()
+                .read()
+        };
+        core::mem::forget(affine_points);
+
+        result
```

Comment:
We could just go the safe route here:

```suggestion
    affine_points.into_iter().map { |point| NonIdentity { point } ).collect() 
```

The `unsafe` usage here seems like a big dance to ensure there isn't a second heap allocation, but is there actually one or is the compiler able to reuse the existing one with safe code?

Reply:
Done, let me know if you want me remove the `debug_assert!`s.